ONSdigital/census-rh-performance-test
# Cr 899 - Add Locust Tests to Request SMS and Postal Fulfilments

# Motivation and Context

Locust tests are required that navigate to the RHUI page for requesting both SMS and postal fulfilments for a Household and to request them both.

This is in order to build out the variety of journeys used in locust so that we can get a better view of performance while RH is being exercised through multiple workflows, and this journey will also need to be used in isolation to measure the performance of RH fulfilment functionality, before and after we have added the fulfilment abuse limiting to measure the impact of our implementation.

# What has changed

Two task sequence classes have been created, one to test the SMS fulfilment requests, and the other to test the postal fulfilment requests. These are named:
- class RequestNewCodeSMS(SequentialTaskSet)
- class RequestNewCodePost(SequentialTaskSet):

New Page objects have been added so that tasks can verify that expected text exists, between the start and end text located, on the pages that get opened:
SELECT_ADDRESS = ('<title>Select your address - Census 2021</title>',
                      '<h1 class="question__title">Select your address</h1>',
                      'I cannot find my address')
SELECT_METHOD = ('<title>How would you like to receive a new access code? - Census 2021</title>',
                    '<h1 class="question__title">How would you like to receive a new household access code?</h1>',
                    'To request a census in a different format or for further help, please')
ENTER_MOBILE = ('<title>What is your mobile phone number? - Census 2021</title>',
                '<h1 class="question__title">What is your mobile phone number?</h1>',
                'Continue')
CONFIRM_MOBILE = ('<title>Is this mobile phone number correct? - Census 2021</title>',
                    '<h1 class="question__title">Is this mobile phone number correct?</h1>',
                    'Continue')
ENTER_NAME = ('<title>What is your name? - Census 2021</title>',
                '<h1 class="question__title">What is your name?</h1>',
                'Continue')
CONFIRM_NAME = ('<title>Do you want to send a new access code to this address? - Census 2021</title>',
                '<h1 class="question__title">Do you want to send a new household access code to this address?</h1>',
                'Continue')
CODE_SENT = ('<title>We have sent an access code - Census 2021</title>',
             '',
             '')

A method named extractAddress has been added to the WebsiteUser class, which extracts the uprn and address, in JSON format, from the HTML in the 'SELECT_ADDRESS' page. The JSON is then used in the post request, on that page, which should cause the ADDRESS_CORRECT page to open.




# How to test?

To test the changes run locust as described in the README. However, please note, you will need to amend the locustfile to point to the task(s) required:

In class WebsiteUser (around line 379) amend the following so that the required tasks have a 1 next to them (and the others have 0 next to them) E.g.:
tasks = {
LaunchEQ: 0,
LaunchEQInvalidUAC: 0,
LaunchEQwithAddressCorrection: 0,
RequestNewCodeSMS: 0,
RequestNewCodePost: 1,
LaunchWebChat: 0
}

For more information about running Locust tasks locally please see this article:
https://collaborate2.ons.gov.uk/confluence/display/SDC/How+to+Run+Locust+Tests+Locally
